app/src/main/cpp/skyline/services/codec/IHardwareOpusDecoder.cpp

```diff
@@ -0,0 +1,96 @@
+    }
+
+    IHardwareOpusDecoder::~IHardwareOpusDecoder() {
+        opus_decoder_destroy(decoderState);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        return DecodeInterleavedImpl(request, response);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        return DecodeInterleavedImpl(request, response, true);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfAndResetOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        return DecodeInterleaved(session, request, response);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleaved(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto reset{request.Pop<bool>()};
+        if (reset)
+            ResetContext();
+
+        return DecodeInterleavedImpl(request, response, true);
+    }
+
+    void IHardwareOpusDecoder::ResetContext() {
+        opus_decoder_ctl(decoderState, OPUS_RESET_STATE);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedImpl(ipc::IpcRequest &request, ipc::IpcResponse &response, bool performanceInfo) {
+        auto dataIn{request.inputBuf.at(0)};
+        auto dataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        if (dataIn.size() <= sizeof(OpusDataHeader))
+            return result::BadOpusDataHeader;
+
+        opus_int32 opusPacketSize{static_cast<opus_int32>(util::SwapEndianness(dataIn.as<OpusDataHeader>().size))};
+        i32 requiredInSize{static_cast<i32>(opusPacketSize + sizeof(OpusDataHeader))};
+        if (opusPacketSize > MaxInputBufferSize || dataIn.size() < requiredInSize)
+            return result::OpusBufferTooSmall;
+
+        // Subtract header from the input buffer to get the opus packet
+        auto sampleDataIn = dataIn.subspan(sizeof(OpusDataHeader));
+
+        auto perfTimer = timesrv::TimeSpanType::FromNanoseconds(util::GetTimeNs());
+        i32 decodedCount{opus_decode(decoderState, sampleDataIn.data(), opusPacketSize, dataOut.data(), decoderOutputBufferSize, false)};
+        perfTimer = timesrv::TimeSpanType::FromNanoseconds(util::GetTimeNs()) - perfTimer;
+
+        if (decodedCount < 0) {
+            switch (decodedCount) {
+                case OPUS_BAD_ARG:
+                    return result::OpusBadArgument;
+                case OPUS_BUFFER_TOO_SMALL:
+                    return result::OpusBufferTooSmall;
+                case OPUS_INTERNAL_ERROR:
+                    return result::OpusInternalError;
+                case OPUS_INVALID_PACKET:
+                    return result::OpusInvalidPacket;
+                case OPUS_INVALID_STATE:
+                    return result::OpusInvalidState;
+                case OPUS_UNIMPLEMENTED:
+                    return result::OpusUnimplemented;
```

Comment:
These results should never be returned to hos, all should throw an exception containing the result value

Reply:
Also add a comment detailing wjy

Reply:
Should I get rid of that switch case and just throw OpusException? It does handle error codes and error strings by itself so that switch case would be redundant

Reply:
Uh you also need to catch the invalid size results as well but sure as long as you do that